### DISCUSSION TITLE
Creating a dynamic table

### DISCUSSION THREAD

**Discussion by cirillojon** [Budibase/budibase#5640] [Help]
I am trying implement an order form where a user can input their own information while using the form and add new rows as they see fit without pulling data from a data provider. Currently we have a solution in a different workflow, but would like to implement it using budibase. Is there a functionality that can help us accomplish this?

![gif](https://user-images.githubusercontent.com/104590083/165825724-ac2130c6-f481-416a-8d3a-1b1350382f75.gif)

**Comment by melohagan** [comment]
HI @cirillojon 

Thanks for including the gif! :)

If the maximum number of inputs is fixed, then you could try using Form Steps: https://docs.budibase.com/docs/form-steps

![form_steps](https://user-images.githubusercontent.com/101575380/165982999-63dee15a-8d55-401f-9ead-41069d7a4cef.gif)

----

This request has also shown up in this discussion: https://github.com/Budibase/budibase/discussions/5189

I think this could be solved if you could use App State as a source for a data provider, but the dev team may have other suggestions.

I've create a feature request here: https://github.com/Budibase/budibase/issues/5658

**Reply by cirillojon** [reply]
Thanks for the response and the gif!

Trying to implement what you did, but having trouble when step 1 disappears after we arrive in step 2. Any ideas?

**Reply by melohagan** [reply]
Hey @cirillojon It's a bit of a cheat in that Step 2 has two inputs, and Step 3 has three inputs.

**Reply by encikbaek** [reply]
Hi, Can show how you create the step form to perform as that.

**Comment by melohagan** [comment]
Hi @encikbaek 

Here's roughly how's you'd go about using form steps to do this. As mentioned, it is limited but here's how it works:

1. Add a form with a step, container, your form fields and two buttons (+,-)
   ![Screenshot 2022-06-28 at 15 16 04](https://user-images.githubusercontent.com/101575380/176201470-9d430b6a-31b4-4bee-937c-42eda5b389bb.png)
   Disable the minus button for the first row. Change all of the field names to have '1' on the end. E.g.
    ![Screenshot 2022-06-28 at 15 17 09](https://user-images.githubusercontent.com/101575380/176201744-fd8e309e-9f9f-4b2e-bbea-1e0ea8037c82.png)

2. Set the actions of the plus and minus buttons. Go to next step and go to previous step.

3. Duplicate step 1, and rename the copy to step 2.  Then duplicate the container. Change the copy field names to end with '2'. Disable both of the buttons in the first container, and enable both buttons in the second container.
    ![Screenshot 2022-06-28 at 15 20 25](https://user-images.githubusercontent.com/101575380/176202538-d2c8ce89-f667-46ad-91d8-3eee82306947.png)

4. Duplicate Step 2, and rename the copy to step 3. Duplicate one of the containers and end the field names with '3'. Disable the plus button of the third container - there are a max of three dynamic inputs in this example. Disable both buttons in the second container.

5. Add a save button to the form. *Define actions* and add three *Save row* actions. You'll need to save each set of indexed fields:
   ![Screenshot 2022-06-28 at 15 24 19](https://user-images.githubusercontent.com/101575380/176203600-a1684e0c-6153-454c-9c77-4c7375210500.png)
 
    It would also likely be a good idea to add form validation to this as well: https://docs.budibase.com/docs/form-validation

-----

As you can see, the growth rate of the number of containers is quite large as the maximum number of dynamic inputs increase. So it's not ideal for large numbers for sure, but it might help in the meantime!

**Reply by encikbaek** [reply]
hi @melohagan thanks for your sharing. I tried it and its works. thank you..

**Reply by sergioponguta** [reply]
Hi @melohagan is this still the best way to do it?

**Reply by melohagan** [reply]
Hey @sergioponguta I'm not sure if it's the best way, but might be reasonable if the max number of additional inputs is small. 

An alternative may be to build a custom component, although in this case it would be quite a tricky one build I think: 
https://docs.budibase.com/docs/custom-component
